sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -507,43 +508,41 @@ cdef float compute_gradient_positive(float[:,:] val_P,
     cdef:
         int ax
         long i, j, k
-        long K = neighbors.shape[1]
-        long n = val_P.shape[0]
-        float[3] buff
-        float D, Q, pij
+        long n_samples = indptr.shape[0] - 1
+        float dij, qij, pij
         float C = 0.0
         float exponent = (dof + 1.0) / -2.0
     cdef clock_t t1, t2
+    cdef float* buff = <float*> malloc(sizeof(float) * n_dimensions)
```

Comment:
It can also be in dimension 1 (and their is a test enforcing that).  
So now TSNE raise a `ValueError` when using `barnes_hut` and `n_components > 3`